source/common/grpc/google_async_client_cache.h

```diff
@@ -0,0 +1,67 @@
+#include <memory>
+
+#include "envoy/grpc/async_client_manager.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/thread_local/thread_local.h"
+
+namespace Envoy {
+namespace Grpc {
+
+// The RawAsyncClient client cache for Google grpc so channel is not created
+// for each request.
+// TODO(fpliu233): The cache will cause resource leak that a new channel is
```

Comment:
This seems concerning and maybe needs to be fixed in this PR. Ideas:
 * is there a simple way to purge the cache on config push?
 * can we timestamp cache entries and do a periodic purging of ones not used for some interval?
 * can we keep these in an LRU and only hold the most recently accessed K entries? FWIW there's a simple LRU implementation (not generic) in Envoy code already here: https://github.com/envoyproxy/envoy/blob/bd73f3c4da0efffb2593d7c9ecf87788856dc052/source/common/stats/recent_lookups.cc#L13

Reply:
Or shall we just make it to live with the callback created with the config, and not use the singleton?

Reply:
If you can solve the first-request-latency problem without a singleton that'd be perfect.

Reply:
I did the load test, the result looks we can do it without singleton. The tls only lives in each callback, I think it'll get destroyed once the callback is destroyed?

Reply:
@htuch Shall we just keep the AsyncClientCache share pointer in the callback, and not use singleton? Is the callback lifetime long enough to hold it?
So in config.cc we can do this without AsyncClientCacheSingleton
```
    const uint32_t timeout_ms =
        PROTOBUF_GET_MS_OR_DEFAULT(proto_config.grpc_service(), timeout, DefaultTimeout);
    auto async_client_cache =
        std::make_shared<Envoy::Grpc::AsyncClientCache>(
            context.clusterManager().grpcAsyncClientManager(), context.scope(),
            context.threadLocal());
    async_client_cache->init(proto_config.grpc_service());
    callback = [async_client_cache, filter_config, timeout_ms, proto_config,
                transport_api_version = proto_config.transport_api_version()](
                   Http::FilterChainFactoryCallbacks& callbacks) {
      auto client = std::make_unique<Filters::Common::ExtAuthz::GrpcClientImpl>(
          async_client_cache->getAsyncClient(),
          std::chrono::milliseconds(timeout_ms), transport_api_version);
      callbacks.addStreamDecoderFilter(Http::StreamDecoderFilterSharedPtr{
          std::make_shared<Filter>(filter_config, std::move(client))});
    };
```

Reply:
If the callback has similar lifetime to the config then I think it's OK. I don't have enough of this lifecycle internalized yet to know if that's the case.

If it's scoped to the request then I would have expected it would have not looked good in your load test.

Reply:
It looks if we want to make it abstracted, as Harvey suggested in a previous comment, then it might just live in some where outside the filter callback, then a singleton might be needed? Anyway, I'll modify it to the callback function without singleton anyway. I don't have a clear idea of how to abstract it yet.

Reply:
Just to make sure we're on the same page: is the plan to get this working for ext_authz in the most expedient way, and then generalize to other grpc uses as a follow-up?

I'm still not clear on what the callback lifetime is.

Reply:
Yes. Thanks for helping clearing the goal. Change is made

Reply:
Done